engine/src/main/java/org/kigalisim/engine/SingleThreadEngine.java

```diff
@@ -0,0 +1,494 @@
+ */
+public class SingleThreadEngine implements Engine {
+
+  private static final Set<String> STREAM_NAMES = new HashSet<>();
+  private static final boolean OPTIMIZE_RECALCS = true;
+
+  static {
+    STREAM_NAMES.add("priorEquipment");
+    STREAM_NAMES.add("equipment");
+    STREAM_NAMES.add("export");
+    STREAM_NAMES.add("import");
+    STREAM_NAMES.add("manufacture");
+    STREAM_NAMES.add("sales");
+  }
+
+  private final int startYear;
+  private final int endYear;
+  private int currentYear;
+
+  private final ConverterStateGetter stateGetter;
+  private final UnitConverter unitConverter;
+  private final StreamKeeper streamKeeper;
+  private Scope scope;
+
+  /**
+   * Create a new SingleThreadEngine instance.
+   *
+   * @param startYear The starting year of the simulation
+   * @param endYear The ending year of the simulation
+   */
+  public SingleThreadEngine(int startYear, int endYear) {
+    // Ensure start year is less than or equal to end year
+    int startYearRearrange = Math.min(startYear, endYear);
+    int endYearRearrange = Math.max(startYear, endYear);
+
+    this.startYear = startYearRearrange;
+    this.endYear = endYearRearrange;
+    this.currentYear = this.startYear;
+
+    this.stateGetter = new ConverterStateGetter(this);
+    this.unitConverter = new UnitConverter(this.stateGetter);
+    this.streamKeeper = new StreamKeeper(
+        new OverridingConverterStateGetter(this.stateGetter), this.unitConverter);
+    this.scope = new Scope(null, null, null);
+  }
+
+  @Override
+  public int getStartYear() {
+    return startYear;
+  }
+
+  @Override
+  public int getEndYear() {
+    return endYear;
+  }
+
+  @Override
+  public void setStanza(String newStanza) {
+    this.scope = this.scope.getWithStanza(newStanza);
+  }
+
+  @Override
+  public void setApplication(String newApplication) {
+    this.scope = this.scope.getWithApplication(newApplication);
+  }
+
+  @Override
+  public void setSubstance(String newSubstance, Boolean checkValid) {
+    this.scope = this.scope.getWithSubstance(newSubstance);
+
+    boolean checkValidEffective = checkValid != null && checkValid;
+    String application = this.scope.getApplication();
+
+    if (checkValidEffective) {
+      boolean knownSubstance = this.streamKeeper.hasSubstance(application, newSubstance);
+      if (!knownSubstance) {
+        throw new RuntimeException("Tried accessing unknown app / substance pair: "
+            + application + ", " + newSubstance);
+      }
+    } else {
+      this.streamKeeper.ensureSubstance(application, newSubstance);
+    }
+  }
+
+  @Override
+  public void setSubstance(String newSubstance) {
+    setSubstance(newSubstance, false);
+  }
+
+  @Override
+  public Scope getScope() {
+    return scope;
+  }
+
+  @Override
+  public void incrementYear() {
+    if (getIsDone()) {
+      throw new RuntimeException("Already completed.");
+    }
+    this.currentYear += 1;
```

Comment:
Please try to not use this. except when in a setter and the parameter has the same name. Please check for similar instances across the file.